drivers/sdhc/sdhc_renesas_ra.c

```diff
@@ -0,0 +1,736 @@
+	fsp_err = R_SDHI_StatusGet(&priv->sdmmc_ctrl, &status);
+	ret = err_fsp2zep(fsp_err);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return (status.card_inserted);
+}
+
+static int sdhc_ra_card_busy(const struct device *dev)
+{
+	struct sdhc_ra_priv *priv = dev->data;
+	fsp_err_t fsp_err;
+	int ret;
+	sdmmc_status_t status;
+
+	fsp_err = R_SDHI_StatusGet(&priv->sdmmc_ctrl, &status);
+	ret = err_fsp2zep(fsp_err);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return (status.transfer_in_progress);
+}
+
+static int sdhi_command_send_wait(sdhi_instance_ctrl_t *p_ctrl, uint32_t command, uint32_t argument,
+				  uint32_t timeout)
+{
+	/* Verify the device is not busy. */
+	r_sdhi_wait_for_device(p_ctrl);
+
+	/* Send the command. */
+	r_sdhi_command_send_no_wait(p_ctrl, command, argument);
+
+	/* Wait for end of response, error or timeout */
+	return r_sdhi_wait_for_event(p_ctrl, SDHI_PRV_RESPONSE_BIT, timeout);
+}
+
+static int sdhc_ra_send_cmd(struct sdhc_ra_priv *priv, struct sdmmc_ra_command *ra_cmd, int retries)
+{
+	int fsp_err = 0;
+
+	while (retries > 0) {
+		fsp_err = sdhi_command_send_wait(&priv->sdmmc_ctrl, ra_cmd->opcode, ra_cmd->arg,
+						 ra_cmd->timeout_ms);
+		if (fsp_err != 0) {
+			retries--; /* error, retry */
+		} else {
+			break;
+		}
+	}
+	return err_fsp2zep(fsp_err);
+}
+
+/*
+ * Send CMD or CMD/DATA via SDHC
+ */
+static int sdhc_ra_request(const struct device *dev, struct sdhc_command *cmd,
+			   struct sdhc_data *data)
+{
+	struct sdhc_ra_priv *priv = dev->data;
+	int retries = (int)(cmd->retries + 1); /* first try plus retries */
+	uint32_t timeout_cfg = 0;
+	fsp_err_t fsp_err = 0;
+	int ret = 0;
+	sdmmc_priv_csd_reg_t p_csd_reg;
+
+	struct sdmmc_ra_command ra_cmd = {
+		.opcode = cmd->opcode,
+		.arg = cmd->arg,
+	};
+
+	if (data) {
+		ra_cmd.data = (uint8_t *)data->data;
+		ra_cmd.sector_count = data->blocks;
+		timeout_cfg = data->timeout_ms;
+	} else {
+		timeout_cfg = cmd->timeout_ms;
+	}
+
+	if (cmd->timeout_ms == SDHC_TIMEOUT_FOREVER) {
+		ra_cmd.timeout_ms = SDHI_TIME_OUT_MAX;
+	} else {
+		ra_cmd.timeout_ms = timeout_cfg;
+	}
+
+	/* Reset semaphore */
+	k_sem_reset(&sdmmc_event[priv->channel].transfer_sem);
+	k_sem_take(&priv->thread_lock, K_MSEC(ra_cmd.timeout_ms));
+
+	/*
+	 * Handle opcode with RA specifics
+	 */
+	switch (cmd->opcode) {
+	case SD_GO_IDLE_STATE:
+	case SD_ALL_SEND_CID:
+	case SD_SEND_RELATIVE_ADDR:
+	case SD_SELECT_CARD:
+	case SD_SEND_IF_COND:
+	case SD_SET_BLOCK_SIZE:
+	case SD_ERASE_BLOCK_START:
+	case SD_ERASE_BLOCK_END:
+	case SD_ERASE_BLOCK_OPERATION:
+	case SD_APP_CMD:
+	case SD_SEND_STATUS:
+		/* Send command with argument */
+		ret = sdhc_ra_send_cmd(priv, &ra_cmd, retries);
+		if (ret < 0) {
+			goto end;
+		}
+		break;
+	case SD_SEND_CSD:
+		/* Read card specific data register */
+		ret = sdhc_ra_send_cmd(priv, &ra_cmd, retries);
+		if (ret < 0) {
+			goto end;
+		}
+		/* SDResponseR2 are bits from 8-127, first 8 MSBs are reserved */
+		p_csd_reg.reg.sdrsp10 = priv->sdmmc_ctrl.p_reg->SD_RSP10;
```

Comment:
So why do we need additional handling of the CSD here? Typically the SD host controller does not care which command is being sent, beyond the type (IE if the command contains data or not)